File: Sources/FastReadCore/RSVPEngine.swift
import Foundation
import NaturalLanguage

public enum RSVPEngine {
    /// Bump when the tokenizer or shaper produces a materially different
    /// chunking for the same input. Persisted tokens whose recorded
    /// `tokenizerVersion` doesn't match are discarded and recomputed.
    // Constants & rule tables are single-sourced from Tools/rsvp-spec.json via
    // the generated RSVPSpec (see RSVPSpec.generated.swift). Don't inline new
    // literals here — add them to the spec so all three platforms stay in sync.
    public static let version: Int = RSVPSpec.version
    public static let defaultWPM: Double = RSVPSpec.WPM.default
    public static let minimumSafeWPM: Double = RSVPSpec.WPM.minimumSafe
    /// Latin-script floor for the user-facing WPM slider. CJK content uses the
    /// lower `minimumUserWPM(forCJK:)` floor.
    public static let minimumUserWPM: Double = RSVPSpec.WPM.minimumUserLatin
    public static let maximumWPM: Double = RSVPSpec.WPM.maximum
    public static let wpmStep: Double = RSVPSpec.WPM.step

    /// Script-dependent floor for the WPM slider: CJK is denser per glyph and
    /// already carries a duration multiplier, so it reads comfortably slower.
    public static func minimumUserWPM(forCJK isCJK: Bool) -> Double {
        isCJK ? RSVPSpec.WPM.minimumUserCJK : RSVPSpec.WPM.minimumUserLatin
    }

    public struct FocusSplit: Equatable {
        public let before: String
        public let focus: String
        public let after: String

        public init(before: String, focus: String, after: String) {
            self.before = before
            self.focus = focus
            self.after = after
        }
    }

    public struct ContextWindow: Equatable {
        public let lowerBound: Int
        public let upperBound: Int
        public let hasLeadingOverflow: Bool
        public let hasTrailingOverflow: Bool

        public init(lowerBound: Int, upperBound: Int, hasLeadingOverflow: Bool, hasTrailingOverflow: Bool) {
            self.lowerBound = lowerBound
            self.upperBound = upperBound
            self.hasLeadingOverflow = hasLeadingOverflow
            self.hasTrailingOverflow = hasTrailingOverflow
        }

        public var range: Range<Int> {
            lowerBound..<upperBound
        }

        public var count: Int {
            max(upperBound - lowerBound, 0)
        }
    }

    public static func countReadingUnits(_ input: String?) -> Int {
        tokenize(input).filter { token in
            token.contains { isFocusable($0) || isCJKCharacter($0) }
        }.count
    }

    public static func containsCJK(_ input: String?) -> Bool {
        guard let input else { return false }
        return hasCJK(input)
    }

    public static func tokenize(_ input: String?, userDictionary: [String] = []) -> [String] {
        guard let input else { return [] }
        let normalized = input.replacingOccurrences(of: "\u{00a0}", with: " ")
        guard hasCJK(normalized) else {
            return normalized
                .split(whereSeparator: { $0.isWhitespace })
                .map(String.init)
        }
        return tokenizeMixed(normalized, userDictionary: userDictionary)
    }

    public static func focusIndex(in token: String?) -> Int {
        let characters = Array(token ?? "")
        guard !characters.isEmpty else { return 0 }

        let focusable = characters.indices.filter { isFocusable(characters[$0]) }
        guard !focusable.isEmpty else {
            return characters.count / 2
        }

        let target: Int
        switch focusable.count {
        case ...1:
            target = 0
        case 2...5:
            target = 1
        case 6...9:
            target = 2
        case 10...13:
            target = 3
        default:
            target = 4
        }

        return focusable[min(target, focusable.count - 1)]
    }

    public static func splitForFocus(_ token: String?) -> FocusSplit {
        let characters = Array(token ?? "")
        guard !characters.isEmpty else {
            return FocusSplit(before: "", focus: "", after: "")
        }

        let index = focusIndex(in: token)
        return FocusSplit(
            before: String(characters[..<index]),
            focus: String(characters[index]),
            after: String(characters[(index + 1)...])
        )
    }

    public static func duration(for token: String?, wpm: Double, punctuationPause: Bool = true) -> Int {
        let safeWPM = safeWPM(wpm)
        let base = RSVPSpec.Duration.baseMillisPerMinute / safeWPM
        let rawToken = token ?? ""
        let isCJK = hasCJK(rawToken)
        let focusableCount = Array(rawToken).filter { isFocusable($0) || isCJKCharacter($0) }.count
        let cjkMultiplier = isCJK ? RSVPSpec.Duration.cjkMultiplier : 1
        let threshold = RSVPSpec.Duration.lengthThreshold
        let lengthMultiplier = focusableCount > threshold
            ? 1 + min(Double(focusableCount - threshold) * RSVPSpec.Duration.lengthPerCharIncrement, RSVPSpec.Duration.lengthMaxIncrement)
            : 1
        let pauseMultiplier = punctuationPause && endsWithPunctuationPause(rawToken) ? RSVPSpec.Duration.pauseMultiplier : 1

        return Int((base * cjkMultiplier * lengthMultiplier * pauseMultiplier).rounded())
    }

    public static func estimateMinutes(wordCount: Int, wpm: Double) -> Double {
        let safeWPM = safeWPM(wpm)
        return Double(wordCount) / safeWPM
    }

    public static func contextWindow(
        tokenCount: Int,
        currentIndex: Int,
        before: Int = 18,
        after: Int = 36
    ) -> ContextWindow {
        guard tokenCount > 0 else {
            return ContextWindow(
                lowerBound: 0,
                upperBound: 0,
                hasLeadingOverflow: false,
                hasTrailingOverflow: false
            )
        }

        let safeBefore = max(before, 0)
        let safeAfter = max(after, 0)
        let clampedIndex = clamp(currentIndex, min: 0, max: tokenCount - 1)
        let lowerBound = max(0, clampedIndex - safeBefore)
        let upperBound = min(tokenCount, clampedIndex + safeAfter + 1)
        return ContextWindow(
            lowerBound: lowerBound,
            upperBound: upperBound,
            hasLeadingOverflow: lowerBound > 0,
            hasTrailingOverflow: upperBound < tokenCount
        )
    }

    public static func clamp<T: Comparable>(_ value: T, min: T, max: T) -> T {
        Swift.min(Swift.max(value, min), max)
    }

    private static func isFocusable(_ character: Character) -> Bool {
        character.unicodeScalars.contains { scalar in
            switch scalar.properties.generalCategory {
            case .uppercaseLetter,
                 .lowercaseLetter,
                 .titlecaseLetter,
                 .modifierLetter,
                 .otherLetter,
                 .decimalNumber,
                 .letterNumber,
                 .otherNumber:
                return true
            default:
                return false
            }
        }
    }

    private static func tokenizeMixed(_ input: String, userDictionary: [String]) -> [String] {
        var tokens: [String] = []
        for segment in splitByScript(input) {
            if segment.isCJK {
                tokens.append(contentsOf: tokenizeCJKSegment(segment.text, userDictionary: userDictionary))
            } else {
                tokens.append(
                    contentsOf: segment.text
                        .split(whereSeparator: { $0.isWhitespace })
                        .map(String.init)
                )
            }
        }
        // ICU + script-split alone over-segments — names, particles, and
        // number+measure-word pairs come out as single Han chars. The
        // ChunkShaper post-pass merges those back into rhythmic 2–4 char
        // chunks that match how the eye-brain pipeline actually reads.
        // See Tools/FastReadTokenizerStats + Tests/Fixtures/rsvp-gold-corpus.tsv.
        return ChunkShaper.shape(tokens)
    }

    private static func splitByScript(_ input: String) -> [(text: String, isCJK: Bool)] {
        var segments: [(String, Bool)] = []
        var buffer = ""
        var bufferIsCJK: Bool? = nil
        for character in input {
            let cjk = isCJKCharacter(character) || isCJKPunctuation(character)
            if let previous = bufferIsCJK, previous != cjk {
                segments.append((buffer, previous))
                buffer = ""
            }
            buffer.append(character)
            bufferIsCJK = cjk
        }
        if let last = bufferIsCJK, !buffer.isEmpty {
            segments.append((buffer, last))
        }
        return segments
    }

    private static func tokenizeCJKSegment(_ text: String, userDictionary: [String]) -> [String] {
        let tokenizer = NLTokenizer(unit: .word)
        tokenizer.string = text

        var raw: [(text: String, lo: String.Index, hi: String.Index)] = []
        tokenizer.enumerateTokens(in: text.startIndex..<text.endIndex) { range, _ in
            let slice = String(text[range])
            if !slice.isEmpty {
                raw.append((slice, range.lowerBound, range.upperBound))
            }
            return true
        }

        let merged = applyUserDictionary(raw, in: text, dictionary: userDictionary)

        var tokens: [String] = []
        // Punctuation that appears in the gap BEFORE the first token (e.g.,
        // an opening 「) is buffered and prepended to the first word, so
        // quoted phrases keep their opening mark instead of dropping it.
        var leadingPunct = ""
        var lastEnd = text.startIndex
        for span in merged {
            if lastEnd < span.lo {
                let gap = text[lastEnd..<span.lo]
                for character in gap {
                    guard isCJKPunctuation(character) || isASCIIPunctuationPause(character) else { continue }
                    if tokens.isEmpty {
                        leadingPunct.append(character)
                    } else {
                        tokens[tokens.count - 1].append(character)
                    }
                }
            }
            if tokens.isEmpty, !leadingPunct.isEmpty {
                tokens.append(leadingPunct + span.text)
                leadingPunct = ""
            } else {
                tokens.append(span.text)
            }
            lastEnd = span.hi
        }
        if lastEnd < text.endIndex {
            attachInterstitialPunctuation(text[lastEnd..<text.endIndex], to: &tokens)
        }
        return tokens
    }

    private static func attachInterstitialPunctuation(_ slice: Substring, to tokens: inout [String]) {
        for character in slice {
            guard isCJKPunctuation(character) || isASCIIPunctuationPause(character) else { continue }
            if !tokens.isEmpty {
                tokens[tokens.count - 1].append(character)
            }
        }
    }

    private static func applyUserDictionary(
        _ raw: [(text: String, lo: String.Index, hi: String.Index)],
        in source: String,
        dictionary: [String]
    ) -> [(text: String, lo: String.Index, hi: String.Index)] {
        let entries = Set(dictionary.filter { !$0.isEmpty })
        guard !entries.isEmpty, !raw.isEmpty else { return raw }
        let maxLength = entries.map(\.count).max() ?? 0
        guard maxLength > 0 else { return raw }

        var result: [(text: String, lo: String.Index, hi: String.Index)] = []
        var index = 0
        while index < raw.count {
            var bestEnd: Int? = nil
            var concat = ""
            var probe = index
            while probe < raw.count {
                concat.append(raw[probe].text)
                if concat.count > maxLength { break }
                if entries.contains(concat) {
                    bestEnd = probe
                }
                probe += 1
            }
            if let end = bestEnd {
                let mergedText = String(source[raw[index].lo..<raw[end].hi])
                result.append((mergedText, raw[index].lo, raw[end].hi))
                index = end + 1
            } else {
                result.append(raw[index])
                index += 1
            }
        }
        return result
    }

    private static func hasCJK(_ string: String) -> Bool {
        string.contains(where: isCJKCharacter)
    }

    private static func isCJKCharacter(_ character: Character) -> Bool {
        character.unicodeScalars.contains { scalar in
            RSVPSpec.cjkRanges.contains { $0.contains(scalar.value) }
        }
    }

    private static func isCJKPunctuation(_ character: Character) -> Bool {
        character.unicodeScalars.contains { scalar in
            RSVPSpec.cjkPunctuationRanges.contains { $0.contains(scalar.value) }
        }
    }

    private static func safeWPM(_ wpm: Double) -> Double {
        let raw = wpm.isFinite && wpm != 0 ? wpm : defaultWPM
        return clamp(raw, min: minimumSafeWPM, max: maximumWPM)
    }

    private static func endsWithPunctuationPause(_ token: String) -> Bool {
        if hasCJK(token) {
            guard let last = token.last else { return false }
            return isCJKPunctuation(last) || isASCIIPunctuationPause(last)
        }

        let punctuation = RSVPSpec.latinSentenceEnd
        let trailingClosers = RSVPSpec.latinTrailingClosers
        let characters = Array(token)

        for index in characters.indices.reversed() {
            if punctuation.contains(characters[index]) {
                let tail = characters.dropFirst(index + 1)
                return tail.allSatisfy { trailingClosers.contains($0) }
            }

            if !trailingClosers.contains(characters[index]) {
                return false
            }
        }

        return false
    }

    private static func isASCIIPunctuationPause(_ character: Character) -> Bool {
        RSVPSpec.asciiPause.contains(character)
    }
}

public enum SourceInputClassifier {
    public static func normalizedURLString(from value: String) -> String? {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, !trimmed.contains(where: { $0.isWhitespace }) else { return nil }

        let normalized = trimmed.range(
            of: #"^[a-z][a-z0-9+.-]*://"#,
            options: [.regularExpression, .caseInsensitive]
        ) == nil
            ? "https://\(trimmed)"
            : trimmed

        guard
            let components = URLComponents(string: normalized),
            let scheme = components.scheme?.lowercased(),
            ["http", "https"].contains(scheme),
            let host = components.host,
            isPlausibleHost(host)
        else {
            return nil
        }

        return normalized
    }

    public static func isLikelySingleURL(_ value: String) -> Bool {
        normalizedURLString(from: value) != nil
    }

    private static func isPlausibleHost(_ host: String) -> Bool {
        host == "localhost" || host.contains(".") || host.contains(":")
    }
}

public enum ArticleTextExtractor {
    public static func readableText(from html: String) -> String {
        normalize(stripTags(html))
    }

    public static func extractTitle(from html: String) -> String? {
        guard let range = html.range(of: #"<title[^>]*>(.*?)</title>"#, options: [.regularExpression, .caseInsensitive]) else {
            return nil
        }
        let raw = html[range]
            .replacingOccurrences(of: #"<[^>]+>"#, with: "", options: .regularExpression)
        let title = normalize(stripTags(String(raw)))
        return title.isEmpty ? nil : title
    }

    public static func stripTags(_ html: String) -> String {
        html
            .replacingOccurrences(of: #"<head[\s\S]*?</head>"#, with: " ", options: [.regularExpression, .caseInsensitive])
            .replacingOccurrences(of: #"<(script|style|noscript|svg)[\s\S]*?</\1>"#, with: " ", options: [.regularExpression, .caseInsensitive])
            .replacingOccurrences(of: #"</(p|div|article|section|h[1-6]|li|br)[^>]*>"#, with: "\n", options: [.regularExpression, .caseInsensitive])
            .replacingOccurrences(of: #"<[^>]+>"#, with: " ", options: .regularExpression)
            .replacingOccurrences(of: "&nbsp;", with: " ")
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&#39;", with: "'")
            .replacingOccurrences(of: "&apos;", with: "'")
            .replacingOccurrences(of: "&rsquo;", with: "'")
            .replacingOccurrences(of: "&lsquo;", with: "'")
            .replacingOccurrences(of: "&ldquo;", with: "\"")
            .replacingOccurrences(of: "&rdquo;", with: "\"")
            .replacingOccurrences(of: "&mdash;", with: "-")
            .replacingOccurrences(of: "&ndash;", with: "-")
    }

    public static func normalize(_ text: String) -> String {
        text
            .components(separatedBy: .whitespacesAndNewlines)
            .filter { !$0.isEmpty }
            .joined(separator: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

File: android-spike/app/src/main/java/com/shhtheonlyperson/fastread/spike/core/RSVPEngine.kt
// Direct port of Sources/FastReadCore/RSVPEngine.swift to Kotlin.
// Same hybrid pipeline: script-split → ICU BreakIterator on CJK runs →
// whitespace split on Latin runs → user-dictionary post-merge → glue
// punctuation to the previous token. Behaviour and output should match
// the Swift parity-corpus.json fixture for any input.

package com.shhtheonlyperson.fastread.spike.core

import java.text.BreakIterator
import java.util.Locale

object RSVPEngine {
    /// Bump when the tokenizer or shaper produces a materially different
    /// chunking for the same input. Persistence drops cached tokens whose
    /// stored `tokenizerVersion` no longer matches this value. Kept in
    /// lock-step with Sources/FastReadCore/RSVPEngine.swift.
    // Constants & rule tables are single-sourced from Tools/rsvp-spec.json via
    // the generated RsvpSpec (see RsvpSpec.generated.kt). Don't inline new
    // literals here — add them to the spec so all three platforms stay in sync.
    const val VERSION: Int = RsvpSpec.VERSION
    const val DEFAULT_WPM: Double = RsvpSpec.Wpm.DEFAULT
    const val MINIMUM_SAFE_WPM: Double = RsvpSpec.Wpm.MINIMUM_SAFE
    const val MAXIMUM_WPM: Double = RsvpSpec.Wpm.MAXIMUM

    data class FocusSplit(val before: String, val focus: String, val after: String)

    data class ContextWindow(
        val lowerBound: Int,
        val upperBound: Int,
        val hasLeadingOverflow: Boolean,
        val hasTrailingOverflow: Boolean,
    ) {
        val range: IntRange get() = lowerBound until upperBound
        val count: Int get() = (upperBound - lowerBound).coerceAtLeast(0)
    }

    fun countReadingUnits(input: String?): Int =
        tokenize(input).count { token -> token.any { isFocusable(it) || isCJKCharacter(it) } }

    fun containsCJK(input: String?): Boolean = input != null && hasCJK(input)

    fun tokenize(input: String?, userDictionary: List<String> = emptyList()): List<String> {
        if (input == null) return emptyList()
        val normalized = input.replace(' ', ' ')
        if (!hasCJK(normalized)) {
            return normalized.split(Regex("\\s+")).filter { it.isNotEmpty() }
        }
        return tokenizeMixed(normalized, userDictionary)
    }

    fun focusIndex(token: String?): Int {
        val chars = token?.toCharArray() ?: charArrayOf()
        if (chars.isEmpty()) return 0

        val focusable = chars.indices.filter { isFocusable(chars[it]) }
        if (focusable.isEmpty()) return chars.size / 2

        val target = when (focusable.size) {
            in 0..1 -> 0
            in 2..5 -> 1
            in 6..9 -> 2
            in 10..13 -> 3
            else -> 4
        }
        return focusable[minOf(target, focusable.lastIndex)]
    }

    fun splitForFocus(token: String?): FocusSplit {
        val chars = token?.toCharArray() ?: charArrayOf()
        if (chars.isEmpty()) return FocusSplit("", "", "")
        val idx = focusIndex(token)
        return FocusSplit(
            before = String(chars, 0, idx),
            focus = chars[idx].toString(),
            after = String(chars, idx + 1, chars.size - idx - 1),
        )
    }

    fun duration(token: String?, wpm: Double, punctuationPause: Boolean = true): Int {
        val safeWpm = safeWPM(wpm)
        val base = RsvpSpec.Duration.BASE_MILLIS_PER_MINUTE / safeWpm
        val raw = token ?: ""
        val isCJK = hasCJK(raw)
        val focusableCount = raw.toCharArray().count { isFocusable(it) || isCJKCharacter(it) }
        val cjkMultiplier = if (isCJK) RsvpSpec.Duration.CJK_MULTIPLIER else 1.0
        val threshold = RsvpSpec.Duration.LENGTH_THRESHOLD
        val lengthMultiplier = if (focusableCount > threshold)
            1.0 + minOf((focusableCount - threshold) * RsvpSpec.Duration.LENGTH_PER_CHAR_INCREMENT, RsvpSpec.Duration.LENGTH_MAX_INCREMENT)
        else 1.0
        val pauseMultiplier = if (punctuationPause && endsWithPunctuationPause(raw)) RsvpSpec.Duration.PAUSE_MULTIPLIER else 1.0
        return (base * cjkMultiplier * lengthMultiplier * pauseMultiplier).toInt()
    }

    fun estimateMinutes(wordCount: Int, wpm: Double): Double {
        val safeWpm = safeWPM(wpm)
        return wordCount / safeWpm
    }

    fun contextWindow(
        tokenCount: Int,
        currentIndex: Int,
        before: Int = 18,
        after: Int = 36,
    ): ContextWindow {
        if (tokenCount <= 0) return ContextWindow(0, 0, false, false)
        val safeBefore = before.coerceAtLeast(0)
        val safeAfter = after.coerceAtLeast(0)
        val clampedIndex = currentIndex.coerceIn(0, tokenCount - 1)
        val lower = maxOf(0, clampedIndex - safeBefore)
        val upper = minOf(tokenCount, clampedIndex + safeAfter + 1)
        return ContextWindow(lower, upper, lower > 0, upper < tokenCount)
    }

    // -- internals -----------------------------------------------------

    private fun tokenizeMixed(input: String, userDictionary: List<String>): List<String> {
        val tokens = mutableListOf<String>()
        for ((segment, isCJK) in splitByScript(input)) {
            if (isCJK) {
                tokens += tokenizeCJKSegment(segment, userDictionary)
            } else {
                tokens += segment.split(Regex("\\s+")).filter { it.isNotEmpty() }
            }
        }
        // ICU + script-split alone over-segments — names, particles, and
        // number+measure-word pairs come out as single Han chars. The
        // ChunkShaper post-pass merges those back into rhythmic 2–4 char
        // chunks that match how the eye-brain pipeline actually reads.
        // The Swift side runs the same five passes against the same
        // gold corpus (Tests/Fixtures/rsvp-gold-corpus.tsv).
        return ChunkShaper.shape(tokens)
    }

    private fun splitByScript(input: String): List<Pair<String, Boolean>> {
        val result = mutableListOf<Pair<String, Boolean>>()
        val buffer = StringBuilder()
        var bufferIsCJK: Boolean? = null
        for (ch in input) {
            val cjk = isCJKCharacter(ch) || isCJKPunctuation(ch)
            if (bufferIsCJK != null && bufferIsCJK != cjk) {
                result += buffer.toString() to bufferIsCJK!!
                buffer.setLength(0)
            }
            buffer.append(ch)
            bufferIsCJK = cjk
        }
        if (buffer.isNotEmpty() && bufferIsCJK != null) {
            result += buffer.toString() to bufferIsCJK!!
        }
        return result
    }

    private data class Span(val text: String, val lo: Int, val hi: Int)

    private fun tokenizeCJKSegment(text: String, userDictionary: List<String>): List<String> {
        // Java's BreakIterator on Android (API 24+) is ICU-backed. Apple's
        // NLTokenizer wraps the same ICU dictionary, so word boundaries
        // are effectively identical.
        val iterator = BreakIterator.getWordInstance(detectCJKLocale(text))
        iterator.setText(text)

        val raw = mutableListOf<Span>()
        var start = iterator.first()
        var end = iterator.next()
        while (end != BreakIterator.DONE) {
            val slice = text.substring(start, end)
            // BreakIterator emits whitespace-only / punctuation-only chunks
            // between word tokens. Drop them — punctuation gets glued back
            // at the end, whitespace just disappears.
            if (slice.any { isFocusable(it) || isCJKCharacter(it) }) {
                raw += Span(slice, start, end)
            }
            start = end
            end = iterator.next()
        }

        val merged = applyUserDictionary(raw, text, userDictionary)

        val tokens = mutableListOf<String>()
        // Punctuation in the gap BEFORE the first token (e.g., an opening
        // 「) buffers and prepends to the first word — same fix as the
        // Swift version, so quoted phrases keep their opening mark.
        var leadingPunct = StringBuilder()
        var lastEnd = 0
        for (span in merged) {
            if (lastEnd < span.lo) {
                val gap = text.substring(lastEnd, span.lo)
                for (ch in gap) {
                    if (!(isCJKPunctuation(ch) || isASCIIPunctuationPause(ch))) continue
                    if (tokens.isEmpty()) {
                        leadingPunct.append(ch)
                    } else {
                        tokens[tokens.lastIndex] = tokens.last() + ch
                    }
                }
            }
            if (tokens.isEmpty() && leadingPunct.isNotEmpty()) {
                tokens += leadingPunct.toString() + span.text
                leadingPunct = StringBuilder()
            } else {
                tokens += span.text
            }
            lastEnd = span.hi
        }
        if (lastEnd < text.length) {
            attachInterstitialPunctuation(text.substring(lastEnd, text.length), tokens)
        }
        return tokens
    }

    private fun attachInterstitialPunctuation(slice: String, tokens: MutableList<String>) {
        for (ch in slice) {
            if (!(isCJKPunctuation(ch) || isASCIIPunctuationPause(ch))) continue
            if (tokens.isNotEmpty()) {
                tokens[tokens.lastIndex] = tokens.last() + ch
            }
        }
    }

    private fun applyUserDictionary(
        spans: List<Span>,
        source: String,
        dictionary: List<String>,
    ): List<Span> {
        val entries = dictionary.filter { it.isNotEmpty() }.toSet()
        if (entries.isEmpty() || spans.isEmpty()) return spans
        val maxLen = entries.maxOf { it.length }

        val result = mutableListOf<Span>()
        var index = 0
        while (index < spans.size) {
            var bestEnd: Int? = null
            val concat = StringBuilder()
            var probe = index
            while (probe < spans.size) {
                concat.append(spans[probe].text)
                if (concat.length > maxLen) break
                if (entries.contains(concat.toString())) bestEnd = probe
                probe += 1
            }
            if (bestEnd != null) {
                val first = spans[index]
                val last = spans[bestEnd]
                result += Span(source.substring(first.lo, last.hi), first.lo, last.hi)
                index = bestEnd + 1
            } else {
                result += spans[index]
                index += 1
            }
        }
        return result
    }

    private fun detectCJKLocale(text: String): Locale {
        val scalars = text.codePoints().toArray()
        if (scalars.any { it in 0xAC00..0xD7AF }) return Locale.KOREAN
        if (scalars.any { it in 0x3040..0x30FF }) return Locale.JAPANESE
        return Locale.TRADITIONAL_CHINESE
    }

    private fun hasCJK(s: String): Boolean = s.any(::isCJKCharacter)

    private fun isCJKCharacter(c: Char): Boolean = RsvpSpec.cjkRanges.any { c.code in it }

    private fun isCJKPunctuation(c: Char): Boolean = RsvpSpec.cjkPunctuationRanges.any { c.code in it }

    private fun isFocusable(c: Char): Boolean {
        if (c.isLetter() || c.isDigit()) return true
        return false
    }

    private fun isASCIIPunctuationPause(c: Char): Boolean = c in RsvpSpec.asciiPause

    private fun safeWPM(wpm: Double): Double {
        val raw = if (wpm.isFinite() && wpm != 0.0) wpm else DEFAULT_WPM
        return raw.coerceIn(MINIMUM_SAFE_WPM, MAXIMUM_WPM)
    }

    private fun endsWithPunctuationPause(token: String): Boolean {
        if (token.isEmpty()) return false
        if (hasCJK(token)) {
            val last = token.last()
            return isCJKPunctuation(last) || isASCIIPunctuationPause(last)
        }
        val punctuation = RsvpSpec.latinSentenceEnd
        val trailingClosers = RsvpSpec.latinTrailingClosers
        for (i in token.indices.reversed()) {
            if (punctuation.contains(token[i])) {
                val tail = token.substring(i + 1)
                return tail.all { trailingClosers.contains(it) }
            }
            if (!trailingClosers.contains(token[i])) return false
        }
        return false
    }
}
